### PR TITLE
fix: make catalog-test-upgrade command

### PR DESCRIPTION
**Pull request by mateusoliveira43**
## Why the changes were made

Allow to test upgrades between OADP versions with different Go versions.

## How to test the changes made

Follow [docs](https://github.com/mateusoliveira43/oadp-operator/blob/fix/catalog-test-upgrade-command/docs/developer/testing/test_oadp_version_upgrade.md) and check if everything works.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -434,22 +434,24 @@ catalog-push: ## Push a catalog image.
...
 catalog-test-upgrade: PREVIOUS_OPERATOR_IMAGE?=ttl.sh/oadp-operator-previous-$(GIT_REV):1h
 catalog-test-upgrade: PREVIOUS_BUNDLE_IMAGE?=ttl.sh/oadp-operator-previous-bundle-$(GIT_REV):1h
 catalog-test-upgrade: THIS_OPERATOR_IMAGE?=ttl.sh/oadp-operator-$(GIT_REV):1h
 catalog-test-upgrade: THIS_BUNDLE_IMAGE?=ttl.sh/oadp-operator-bundle-$(GIT_REV):1h
 catalog-test-upgrade: CATALOG_IMAGE?=ttl.sh/oadp-operator-catalog-$(GIT_REV):1h
 catalog-test-upgrade: opm login-required ## Prepare a catalog image with two channels: PREVIOUS_CHANNEL and from current branch
-	cp -r ./ $(TEMP)/current
-	git clone --depth=1 git@github.com:openshift/oadp-operator.git -b $(PREVIOUS_CHANNEL) $(TEMP)/$(PREVIOUS_CHANNEL)
-	cd $(TEMP)/$(PREVIOUS_CHANNEL) && IMG=$(PREVIOUS_OPERATOR_IMAGE) BUNDLE_IMG=$(PREVIOUS_BUNDLE_IMAGE) make bundle && \
+	# mkdir test-upgrade && cp -r ./ test-upgrade/current
```
Comment (mateusoliveira43) [inline Makefile:446]: removed temp to avoid problems, but now I have a can not copy folder into itself problem. Ideas?

File: Makefile [hunk 2]
```diff
@@ -458,7 +460,7 @@ catalog-test-upgrade: opm login-required ## Prepare a catalog image with two cha
 	echo -e "apiVersion: operators.coreos.com/v1alpha1\nkind: CatalogSource\nmetadata:\n  name: oadp-operator-catalog-test-upgrade\n  namespace: openshift-marketplace\nspec:\n  sourceType: grpc\n  image: $(CATALOG_IMAGE)" | $(OC_CLI) create -f -
 	# To delete it afterwards, run `$(OC_CLI) delete catalogsource oadp-operator-catalog-test-upgrade -n openshift-marketplace`
 	# or `make undeploy-olm`
-	chmod -R 777 $(TEMP) && rm -rf $(TEMP)
+	# chmod -R 777 test-upgrade && rm -rf test-upgrade && docker image rm catalog-test-upgrade
```
Comment (mateusoliveira43) [inline Makefile:463]: commented for tests, will uncomment

**Comment by mpryc** [comment]
Remember when you had PR for cloning and running make commands inside temp folder, would it be useful here?

File: Makefile [hunk 3]
```diff
@@ -434,22 +434,24 @@ catalog-push: ## Push a catalog image.
...
 catalog-test-upgrade: THIS_BUNDLE_IMAGE?=ttl.sh/oadp-operator-bundle-$(GIT_REV):1h
 catalog-test-upgrade: CATALOG_IMAGE?=ttl.sh/oadp-operator-catalog-$(GIT_REV):1h
 catalog-test-upgrade: opm login-required ## Prepare a catalog image with two channels: PREVIOUS_CHANNEL and from current branch
-	cp -r ./ $(TEMP)/current
-	git clone --depth=1 git@github.com:openshift/oadp-operator.git -b $(PREVIOUS_CHANNEL) $(TEMP)/$(PREVIOUS_CHANNEL)
-	cd $(TEMP)/$(PREVIOUS_CHANNEL) && IMG=$(PREVIOUS_OPERATOR_IMAGE) BUNDLE_IMG=$(PREVIOUS_BUNDLE_IMAGE) make bundle && \
+	# mkdir test-upgrade && cp -r ./ test-upgrade/current
+	git clone --depth=1 git@github.com:openshift/oadp-operator.git -b $(PREVIOUS_CHANNEL) test-upgrade/$(PREVIOUS_CHANNEL)
+	cd test-upgrade/$(PREVIOUS_CHANNEL) && \
+		echo -e "FROM golang:$(PREVIOUS_CHANNEL_GO_VERSION)\nRUN useradd --create-home dev\nUSER dev\nWORKDIR /home/dev/$(PREVIOUS_CHANNEL)" | docker image build --tag catalog-test-upgrade - && \
```
Comment (mpryc) [inline Makefile:449]: Don't we want to have podman and docker allowed? e.g. on standard Fedora there is no docker CLI unless wrapper it's installed.

Comment (mateusoliveira43) [reply Makefile:449]: currently we do not use podman in our Makefile

so I would say this is another issue, but worth attacking later yes

Comment (mpryc) [reply Makefile:449]: Ok

File: Makefile (hunk 1 above)
Comment (mpryc) [reply Makefile:446]: use rsync, much preferred way, see:
https://github.com/openshift/oadp-operator/pull/1391/files#diff-8de5118a56011d553059dd47da58cea16ecd90598ef82d0d64bcba3ce9cc0423R67

Comment (mateusoliveira43) [reply Makefile:446]: going to check out

**Review by kaovilai — DISMISSED** [review]

File: Makefile (hunk 1 above)
Comment (mateusoliveira43) [reply Makefile]: added rsync

**Review by kaovilai — DISMISSED** [review]

File: Makefile (hunk 3 above)
Comment (mpryc) [reply Makefile:450]: Added in the last commit, two new tests as in the commit message. Thanks.

File: docs/developer/testing/test_oadp_version_upgrade.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+This will create `oadp-operator-catalog-test-upgrade` catalog in your cluster (OperatorHub in the UI).
+
+![catalog in OperatorHUB](../../images/test_oadp_version_upgrade_catalog.png)
+
+The catalog will have two channels:
+
+- the previous version (the released version prior to the version of the branch you are)
+
+    this is defined in `PREVIOUS_CHANNEL` variable in Makefile, and can be cahnged
```
Comment (sseago) [inline docs/developer/testing/test_oadp_version_upgrade.md:17]: nit: s/cahnged/changed/

Comment (mateusoliveira43) [reply docs/developer/testing/test_oadp_version_upgrade.md]: done

**Comment by shubham-pampattiwar** [comment]
Operator upgrade worked successfully for me from the custom catalog (oadp-1.3 -> master stable)
```
shubham@fedora ~/oadp-operator -  (mateus-upgrade) $ oc get po
NAME                                                READY   STATUS    RESTARTS   AGE
node-agent-2ltwm                                    1/1     Running   0          88s
node-agent-vnlsc                                    1/1     Running   0          88s
openshift-adp-controller-manager-85dd4d9f4f-2fjlq   0/1     Running   0          10s   <----- new pod coming up
openshift-adp-controller-manager-bc979f446-twbt4    1/1     Running   0          2m21s  <---- old pod going down
velero-844664dffd-9kqhz                             1/1     Running   0          88s

```

**Review by shubham-pampattiwar — APPROVED** [review]

**Review by sseago — APPROVED** [review]

**Comment by mateusoliveira43** [comment]
/override ci/prow/4.15-e2e-test-aws
restore succeeds, but application not working flake

**Pull request merged**